File: 351003/Pulish/publisher/app/services/mark_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from app.dto.mark import MarkRequestTo, MarkResponseTo
from app.models.mark import Mark
from app.core.exceptions import NotFoundException, AppException
from app.cache.redis_client import cache_get, cache_set, cache_delete


class MarkService:
    def __init__(self, db: Session):
        self.db = db

    def create(self, dto: MarkRequestTo) -> MarkResponseTo:
        mark = Mark(name=dto.name)
        try:
            self.db.add(mark)
            self.db.commit()
            self.db.refresh(mark)
        except IntegrityError:
            self.db.rollback()
            raise AppException(
                "Mark with this name already exists", 40303, 403)
        result = self._to_response(mark)
        cache_set(f"mark:{mark.id}", result.model_dump())
        cache_delete("marks:all")
        return result

    def find_all(self):
        cached = cache_get("marks:all")
        if cached is not None:
            return [MarkResponseTo(**m) for m in cached]
        marks = self.db.query(Mark).all()
        result = [self._to_response(m) for m in marks]
        cache_set("marks:all", [r.model_dump() for r in result])
        return result

    def find_by_id(self, id: int):
        cached = cache_get(f"mark:{id}")
        if cached is not None:
            return MarkResponseTo(**cached)
        mark = self.db.query(Mark).filter(Mark.id == id).first()
        if not mark:
            raise NotFoundException("Mark not found", 40403)
        result = self._to_response(mark)
        cache_set(f"mark:{id}", result.model_dump())
        return result

    def update(self, dto: MarkRequestTo):
        mark = self.db.query(Mark).filter(Mark.id == dto.id).first()
        if not mark:
            raise NotFoundException("Mark not found", 40403)

        mark.name = dto.name
        try:
            self.db.commit()
            self.db.refresh(mark)
        except IntegrityError:
            self.db.rollback()
            raise AppException(
                "Mark with this name already exists", 40303, 403)

        result = self._to_response(mark)
        cache_set(f"mark:{mark.id}", result.model_dump())
        cache_delete("marks:all")
        return result

    def delete(self, id: int):
        mark = self.db.query(Mark).filter(Mark.id == id).first()
        if not mark:
            raise NotFoundException("Mark not found", 40403)
        self.db.delete(mark)
        self.db.commit()
        cache_delete(f"mark:{id}")
        cache_delete("marks:all")

    def _to_response(self, mark: Mark) -> MarkResponseTo:
        return MarkResponseTo(id=mark.id, name=mark.name)

File: 351003/Suchok/publisher/src/main/java/by/liza/app/service/MarkService.java
package by.liza.app.service;

import by.liza.app.dto.request.MarkRequestTo;
import by.liza.app.dto.response.MarkResponseTo;
import by.liza.app.exception.DuplicateEntityException;
import by.liza.app.exception.EntityNotFoundException;
import by.liza.app.mapper.MarkMapper;
import by.liza.app.model.Mark;
import by.liza.app.repository.ArticleRepository;
import by.liza.app.repository.MarkRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class MarkService {

    private final MarkRepository    markRepository;
    private final ArticleRepository articleRepository;
    private final MarkMapper        markMapper;

    @Transactional
    @Caching(
            put  = @CachePut(value = "marks", key = "#result.id"),
            evict = @CacheEvict(value = "marks_all", allEntries = true)
    )
    public MarkResponseTo create(MarkRequestTo requestTo) {
        if (markRepository.existsByName(requestTo.getName())) {
            throw new DuplicateEntityException(
                    "Mark with name '" + requestTo.getName() + "' already exists", 40303);
        }
        Mark mark = markMapper.toEntity(requestTo);
        return markMapper.toResponse(markRepository.save(mark));
    }

    @Cacheable(value = "marks", key = "#id")
    public MarkResponseTo getById(Long id) {
        return markMapper.toResponse(findById(id));
    }

    @Cacheable(value = "marks_all")
    public List<MarkResponseTo> getAll() {
        return markMapper.toResponseList(markRepository.findAll());
    }

    @Transactional
    @Caching(
            put  = @CachePut(value = "marks", key = "#result.id"),
            evict = @CacheEvict(value = "marks_all", allEntries = true)
    )
    public MarkResponseTo update(MarkRequestTo requestTo) {
        if (requestTo.getId() == null) {
            throw new EntityNotFoundException("Mark id must be provided for update", 40003);
        }
        Mark existing = findById(requestTo.getId());
        if (markRepository.existsByNameAndIdNot(requestTo.getName(), requestTo.getId())) {
            throw new DuplicateEntityException(
                    "Mark with name '" + requestTo.getName() + "' already exists", 40303);
        }
        existing.setName(requestTo.getName());
        return markMapper.toResponse(markRepository.save(existing));
    }

    @Transactional
    @Caching(evict = {
            @CacheEvict(value = "marks",     key = "#id"),
            @CacheEvict(value = "marks_all", allEntries = true)
    })
    public void delete(Long id) {
        if (!markRepository.existsById(id)) {
            throw new EntityNotFoundException("Mark with id " + id + " not found", 40403);
        }
        markRepository.deleteById(id);
    }

    public List<MarkResponseTo> getByArticleId(Long articleId) {
        if (!articleRepository.existsById(articleId)) {
            throw new EntityNotFoundException("Article with id " + articleId + " not found", 40402);
        }
        return markMapper.toResponseList(markRepository.findByArticleId(articleId));
    }

    private Mark findById(Long id) {
        return markRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(
                        "Mark with id " + id + " not found", 40403));
    }
}